product_variant_search_by_attribute/models/product.py

```diff
@@ -0,0 +1,111 @@
+
+import itertools
+import logging
+
+from openerp import api, fields, models
+
+
+_logger = logging.getLogger(__name__)
+
+
+class ProductProduct(models.Model):
+    _inherit = 'product.product'
+
+    attribute_str = fields.Char(
+        string='Attribute string', compute='_compute_attribute_str',
+        store=True, index=True,
+        help="Store all attribute values. Required to search variants "
+             "from attributes")
+
+    @api.multi
+    @api.depends('attribute_value_ids', 'active')
+    def _compute_attribute_str(self):
+        for prd in self:
+            if prd.attribute_value_ids:
+                attrs = [x.name.lower() for x in prd.attribute_value_ids]
+                prd.attribute_str = u' '.join(attrs)
+
+    @api.multi
+    def _compute_attrib_str_after_attrib_value_change(self):
+        """ This method is called when you modified attribute value name
+            on impacted products: in this case you may have a huge data volume
+        """
+        sql = ''
+        product_ids = [x.id for x in self]
+        self._cr.execute(self._get_product_info_query(), (tuple(product_ids),))
+        product_infos = self._cr.fetchall()
+        products_map = {x[0]: x[1] for x in product_infos}
+        update_statement = "UPDATE product_product SET attribute_str = " \
+                           "%(info)s WHERE id = %(id)s;"
+        for prd in self:
+            if products_map.get(prd.id):
+                # We do not use previous method compute_attribute_str()
+                # for performance reason:
+                # 30 000 products: 5 seconds here
+                #      or 3 minutes for method above
+                query_params = {'info': products_map[prd.id], 'id': prd.id}
+                sql += self._cr.mogrify(update_statement, query_params)
+        if sql:
+            self._cr.execute(sql)
+            self.env.invalidate_all()
+
+    def _get_product_info_query(self):
+        """ You may customize aggregate string according to your needs """
+        return """
+            SELECT re.prod_id as id, lower(string_agg(pa.name, ' ')) as string
+            FROM product_attribute_value_product_product_rel re
+               LEFT JOIN product_attribute_value pa ON pa.id = re.att_id
+            WHERE re.prod_id in %s
+            GROUP BY 1 """
+
+    def search(self, cr, uid, domain, offset=0, limit=None,
+               order=None, context=None, count=False):
+        _logger.debug('Initial domain search %s' % domain)
+        separator = self.pool['ir.config_parameter'].get_param(
+            cr, uid, 'search.by.attribute.separator', ' ')
+        domain = self.domain_replacement(domain, 'attribute_str', separator)
+        return super(ProductProduct, self).search(
+            cr, uid, domain, offset=offset, limit=limit, order=order,
+            context=context, count=count)
+
+    def domain_replacement(self, domain, field, separator):
```

Comment:
I'm not sure to fully understand the use case of OR expression you describe in case of complex domain, the benefit on simplicity that bring us.